xds/internal/xdsclient/client_new.go

```diff
@@ -54,24 +57,48 @@ const NameForServer = "#server"
 // only when all references are released, and it is safe for the caller to
 // invoke this close function multiple times.
 func New(name string) (XDSClient, func(), error) {
-	return newRefCounted(name, defaultWatchExpiryTimeout, defaultIdleAuthorityDeleteTimeout, backoff.DefaultExponential.Backoff)
+	return newRefCounted(name, defaultWatchExpiryTimeout, defaultIdleChannelExpiryTimeout, backoff.DefaultExponential.Backoff)
 }
 
 // newClientImpl returns a new xdsClient with the given config.
-func newClientImpl(config *bootstrap.Config, watchExpiryTimeout time.Duration, idleAuthorityDeleteTimeout time.Duration, streamBackoff func(int) time.Duration) (*clientImpl, error) {
+func newClientImpl(config *bootstrap.Config, watchExpiryTimeout, idleChannelExpiryTimeout time.Duration, streamBackoff func(int) time.Duration) (*clientImpl, error) {
 	ctx, cancel := context.WithCancel(context.Background())
 	c := &clientImpl{
 		done:               grpcsync.NewEvent(),
+		authorities:        make(map[string]*authority),
 		config:             config,
 		watchExpiryTimeout: watchExpiryTimeout,
 		backoff:            streamBackoff,
 		serializer:         grpcsync.NewCallbackSerializer(ctx),
 		serializerClose:    cancel,
+		transportBuilder:   &grpctransport.Builder{},
 		resourceTypes:      newResourceTypeRegistry(),
-		authorities:        make(map[string]*authority),
-		idleAuthorities:    cache.NewTimeoutCache(idleAuthorityDeleteTimeout),
+		xdsActiveChannels:  make(map[string]*channelState),
+		xdsIdleChannels:    cache.NewTimeoutCache(idleChannelExpiryTimeout),
 	}
 
+	for name, cfg := range config.Authorities() {
+		// If server configs are specified in the authorities map, use that.
+		// Else, use the top-level server configs.
+		serverCfg := config.XDSServers()
+		if len(cfg.XDSServers) >= 1 {
+			serverCfg = cfg.XDSServers
+		}
+		c.authorities[name] = newAuthority(authorityBuildOptions{
+			serverConfigs:    serverCfg,
+			name:             name,
+			serializer:       c.serializer,
+			getChannelForADS: c.getChannelForADS,
+			logPrefix:        clientPrefix(c),
+		})
+	}
+	c.topLevelAuthority = newAuthority(authorityBuildOptions{
+		serverConfigs:    config.XDSServers(),
+		name:             "",
+		serializer:       c.serializer,
+		getChannelForADS: c.getChannelForADS,
+		logPrefix:        clientPrefix(c),
+	})
```

Comment:
Can you explain this else use top level server config for authority, and the top level authority logic. I couldn't find the spec/documentation for this for some reason.

Reply:
Every authority entry in the `authorities` map of the bootstrap configuration specifies a list of server configuration for that authority. But this list can be empty, and if that is the case, the top-level `xds_servers` field is used for that authority.

See: https://github.com/grpc/proposal/blob/master/A47-xds-federation.md#bootstrap-config-changes 

And the last line of comment on the `xds_servers` field inside the `authorities` map.

    // If not specified, the top-level server list is used.

Reply:
Ok. And the top level authority gets used when again (I remember reading it somewhere but don't remember what it does)?

Reply:
Yes, the `topLevelAuthority` field here corresponds to the list of the server configurations in the top-level `xds_servers` field in the bootstrap configuration that gets used under the following cases:
- an old style resource name that does not contain an authority in its name
- a new style resource name that matches to some authority in the `authorities` map, but that authority does not have any server configuration.

Reply:
Ok I see. So top level hits from old style resource names and new style hit authority no server configuration. Thanks.